### PR TITLE
Dashboard Settings: Fix TimeZone dropdown doesn't change the timezone

### LINKED ISSUE #35504: DashboardSettings: TimeZone dropdown doesn't change the timezone
**Doesn't work from dashboard settings:**
1. Open this dashboard https://play.grafana.org/d/l4CyfC6Mz/dashboardsettings-timezone-dropdown-doesnt-change-the-timezone-35504?orgId=1
2. Goto dashboard settings
3. Change the time zone, time zone changes [Ok]
3. Go back to dashboard
4. The time zone change is not reflected in panel or in the time picker [NOT OK]

**Works from time picker:**
1. Open this dashboard https://play.grafana.org/d/l4CyfC6Mz/dashboardsettings-timezone-dropdown-doesnt-change-the-timezone-35504?orgId=1
2. Change the time zone in the time picker, time zone changes [Ok]
3. The time zone changes are reflected in time picker [OK]
4. Goto dashboard settings
5. The time zone changes are reflected in time zone picker [OK]

### PR DESCRIPTION
**Pull request by axelavargas**
**What this PR does / why we need it**:

The time zone change is not reflected in the panel or in the time picker

**Approach**

* Now `dashboard.timezone` is being updated using a redux action, I followed the same approach we do on the `DashNavTimeControls` component, as well I added unit test.

timezone

**Which issue(s) this PR fixes**:

Fixes #35504

**Special notes for your reviewer**:

I noticed that on the current version of the timezone is updated in the dashboard settings or in the time picker it's not required to save changes, therefore if the user refresh the page the timezone is not being updated, is it expected behavior?

### CODE REVIEW

File: public/app/features/dashboard/components/DashboardSettings/GeneralSettings.test.tsx [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        { title: 'B', id: 'B' },
+      ]),
+    }),
+  }));
+
+  jest.mock('app/core/services/context_srv', () => ({
+    contextSrv: {
+      user: { orgId: 1 },
+    },
+  }));
```
Comment (hugohaggmark) [inline public/app/features/dashboard/components/DashboardSettings/GeneralSettings.test.tsx:108]: These repeating mocks look weird?

File: public/app/features/dashboard/components/DashboardSettings/GeneralSettings.tsx [hunk 2]
```diff
@@ -7,17 +8,22 @@ import { DashboardModel } from '../../state/DashboardModel';
...
+export type Props = OwnProps & ConnectedProps<typeof connector>;
+
 const GRAPH_TOOLTIP_OPTIONS = [
   { value: 0, label: 'Default' },
   { value: 1, label: 'Shared crosshair' },
   { value: 2, label: 'Shared Tooltip' },
 ];
 
-export const GeneralSettings: React.FC<Props> = ({ dashboard }) => {
+export const GeneralSettingsUnconnected: React.FC<Props> = ({ dashboard, updateTimeZone }) => {
```
Comment (hugohaggmark) [inline public/app/features/dashboard/components/DashboardSettings/GeneralSettings.tsx:26]: super nit: we've decided to move away from React.FC<Props> in favor of
```suggestion
export function GeneralSettingsUnconnected({ dashboard, updateTimeZone }:Props): JSX.Element {
```

File: public/app/features/dashboard/components/DashboardSettings/GeneralSettings.tsx [hunk 3]
```diff
@@ -127,3 +134,14 @@ export const GeneralSettings: React.FC<Props> = ({ dashboard }) => {
     </div>
   );
 };
+
+const mapDispatchToProps = {
+  updateTimeZone: (timeZone: TimeZone) => {
+    updateTimeZoneForSession(timeZone);
+    getTimeSrv().refreshDashboard();
+  },
```
Comment (hugohaggmark) [inline public/app/features/dashboard/components/DashboardSettings/GeneralSettings.tsx:142]: I think we should create a proper ThunkAction in a `actions.ts` file which is our convention.

File: public/app/features/dashboard/components/DashboardSettings/GeneralSettings.test.tsx [hunk 4]
```diff
@@ -0,0 +1,119 @@
+import React from 'react';
```
Comment (hugohaggmark) [inline public/app/features/dashboard/components/DashboardSettings/GeneralSettings.test.tsx:1]: Great that you're adding tests!

File: public/app/features/dashboard/components/DashboardSettings/GeneralSettings.test.tsx [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+import { render, screen } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+
+import { byRole, byText } from 'testing-library-selector';
```
Comment (hugohaggmark) [inline public/app/features/dashboard/components/DashboardSettings/GeneralSettings.test.tsx:5]: This is an interesting helper library, I'll have to check it out. Ah wait it's @domasx2 lib, awesome!

**Review by hugohaggmark — APPROVED** [review]
You continue to deliver really nice PRs with great descriptions, code and tests! 🎉  This works as expected and I left some small comments for you to consider.

File: public/app/features/dashboard/components/DashboardSettings/GeneralSettings.tsx [hunk 6]
```diff
@@ -127,3 +134,14 @@ export const GeneralSettings: React.FC<Props> = ({ dashboard }) => {
...
+};
+
+const connector = connect(null, mapDispatchToProps);
+
+export const GeneralSettings = connector(GeneralSettingsUnconnected);
```
Comment (torkelo) [inline public/app/features/dashboard/components/DashboardSettings/GeneralSettings.tsx:147]: Don't think we need to use connector & mapDispatchToProps, we can just use the hook useDispatch directly from the component function

Comment (hugohaggmark) [reply public/app/features/dashboard/components/DashboardSettings/GeneralSettings.tsx:147]: This makes it easier to test I guess?

Comment (axelavargas) [reply public/app/features/dashboard/components/DashboardSettings/GeneralSettings.tsx:147]: Yes, exactly, I followed this pattern to make testing easier.

**Review by mckn — APPROVED** [review]
I think this looks really good! Great work @axelavargas! 

I found one small thing that looks weird but I think it exists in main as well so we can fix that in a separate PR.

timezonr

File: public/app/features/dashboard/components/DashboardSettings/GeneralSettings.test.tsx (hunk 1 above)
Comment (axelavargas) [reply public/app/features/dashboard/components/DashboardSettings/GeneralSettings.test.tsx:108]: good catch! (_I just started using vim, this is what happened when one is learning the commands_ 😅 )

**Comment by axelavargas** [comment]
> I think this looks really good! Great work @axelavargas!
> 
> I found one small thing that looks weird but I think it exists in main as well so we can fix that in a separate PR.
> 
> timezonr

Thank you @mckn, for this one should we create a new issue? or do you know if is already created?

File: public/app/features/dashboard/components/DashboardSettings/GeneralSettings.tsx [hunk 7]
```diff
@@ -126,4 +133,15 @@ export const GeneralSettings: React.FC<Props> = ({ dashboard }) => {
       </div>
     </div>
   );
+}
+
+const mapDispatchToProps = {
+  updateTimeZone: (timeZone: TimeZone) => {
```
Comment (jackw) [inline public/app/features/dashboard/components/DashboardSettings/GeneralSettings.tsx:139]: Does this action cause `Error: Actions must be plain objects. Use custom middleware for async actions.` in console?

I think this either needs to be made into a proper ThunkAction as @hugohaggmark is suggesting or pass in the `updateTimeZoneForSession` as a prop and then move the body of this action inside `onTimeZoneChange`.

**Review by jackw — APPROVED** [review]
Tested locally and works as expected. Great work @axelavargas! 👍

File: public/app/features/dashboard/components/DashboardSettings/GeneralSettings.tsx (hunk 3 above)
Comment (axelavargas) [reply public/app/features/dashboard/components/DashboardSettings/GeneralSettings.tsx]: good suggestion @hugohaggmark 🚀  787291e092d04907ebbc3db2a5824b8514f72af0

File: public/app/features/dashboard/components/DashboardSettings/GeneralSettings.tsx (hunk 7 above)
Comment (axelavargas) [reply public/app/features/dashboard/components/DashboardSettings/GeneralSettings.tsx]: Thank you Jack :), indeed it was causing an error in the console, I believed is fixed, I followed Hugo suggestion 787291e092d04907ebbc3db2a5824b8514f72af0

**Review by hugohaggmark — APPROVED** [review]
Looks great and works as expected! No more console error 🎉

**Comment by mckn** [comment]
> > I think this looks really good! Great work @axelavargas!
> > I found one small thing that looks weird but I think it exists in main as well so we can fix that in a separate PR.
> > timezonr
> 
> Thank you @mckn, for this one should we create a new issue? or do you know if is already created?

Not sure, I will see if I can find one.

**Pull request merged**